hy/compiler.py

```diff
@@ -1125,117 +1080,60 @@ def compile_if(self, expression):
     @builds("break")
     @checkargs(0)
     def compile_break_expression(self, expr):
-        ret = ast.Break(lineno=expr.start_line,
-                        col_offset=expr.start_column)
-
-        return ret
+        return asty.Break(expr)
 
     @builds("continue")
     @checkargs(0)
     def compile_continue_expression(self, expr):
-        ret = ast.Continue(lineno=expr.start_line,
-                           col_offset=expr.start_column)
-
-        return ret
+        return asty.Continue(expr)
 
     @builds("assert")
     @checkargs(min=1, max=2)
     def compile_assert_expression(self, expr):
         expr.pop(0)  # assert
-        e = expr.pop(0)
-        if len(expr) == 1:
+        ret = self.compile(expr.pop(0))
+        e = ret.force_expr
+        msg = None
+        if expr:
             msg = self.compile(expr.pop(0)).force_expr
-        else:
-            msg = None
-        ret = self.compile(e)
-        ret += ast.Assert(test=ret.force_expr,
-                          msg=msg,
-                          lineno=e.start_line,
-                          col_offset=e.start_column)
-
-        return ret
+        return ret + asty.Assert(expr, test=e, msg=msg)
 
     @builds("global")
-    @checkargs(min=1)
-    def compile_global_expression(self, expr):
-        expr.pop(0)  # global
-        names = []
-        while len(expr) > 0:
-            identifier = expr.pop(0)
-            name = ast_str(identifier)
-            names.append(name)
-            if not isinstance(identifier, HySymbol):
-                raise HyTypeError(identifier, "(global) arguments must "
-                                  " be Symbols")
-
-        return ast.Global(names=names,
-                          lineno=expr.start_line,
-                          col_offset=expr.start_column)
-
     @builds("nonlocal")
     @checkargs(min=1)
-    def compile_nonlocal_expression(self, expr):
-        if not PY3:
+    def compile_global_or_nonlocal(self, expr):
+        form = expr.pop(0)
+        if form == "nonlocal" and not PY3:
```

Comment:
shouldn't this be in the `@builds_if` decorator?

Reply:
Yeah, good idea.

Reply:
Okay, changed and rebased.